torchvision/transforms/v2/functional/_utils.py

```diff
@@ -1,8 +1,110 @@
+import functools
+import inspect
 from typing import Any
 
 import torch
+from torchvision import datapoints
 from torchvision.datapoints._datapoint import Datapoint
 
 
 def is_simple_tensor(inpt: Any) -> bool:
     return isinstance(inpt, torch.Tensor) and not isinstance(inpt, Datapoint)
+
+
+_KERNEL_REGISTRY = {}
+
+
+def _kernel_wrapper_internal(dispatcher, kernel):
+    dispatcher_params = list(inspect.signature(dispatcher).parameters)[1:]
+    kernel_params = list(inspect.signature(kernel).parameters)[1:]
+
+    needs_args_kwargs_handling = kernel_params != dispatcher_params
+
+    # this avoids converting list -> set at runtime below
+    kernel_params = set(kernel_params)
+
+    @functools.wraps(kernel)
+    def wrapper(inpt, *args, **kwargs):
+        input_type = type(inpt)
+
+        if needs_args_kwargs_handling:
+            # Convert args to kwargs to simplify further processing
+            kwargs.update(dict(zip(dispatcher_params, args)))
+            args = ()
+
+            # drop parameters that are not relevant for the kernel, but have a default value
+            # in the dispatcher
+            for kwarg in kwargs.keys() - kernel_params:
+                del kwargs[kwarg]
+
+            # add parameters that are passed implicitly to the dispatcher as metadata,
+            # but have to be explicit for the kernel
+            for kwarg in input_type.__annotations__.keys() & kernel_params:
+                kwargs[kwarg] = getattr(inpt, kwarg)
```

Comment:
I'm not a die-hard fan of all this magic, but OK (for now...). It's not exposed to users, so we can always go back to this if it becomes too weird.